### PR TITLE
Add a new dashboard widget for the route breakdown

### PR DESCRIPTION
**Pull request by raul-gracia**
## Description
Adds a new dashboard widget for the route breakdown

- Creates a new query object to obtain the route breakdown
- Adds a new widget to the dashboard to display the route breakdown

## Trello Card Link

https://trello.com/c/ZgxQldBl

### CODE REVIEW

**Comment by raul-gracia** [comment]
[KPI Widget: Route Breakdown](https://trello.com/c/ZgxQldBl/97-kpi-widget-route-breakdown)

File: .rubocop.yml [hunk 1]
```diff
@@ -43,6 +43,9 @@ RSpec/LeakyConstantDeclaration:
   Exclude:
     - 'spec/**/*'
 
+RSpec/MultipleExpectations:
```
Comment (pmanrubia) [inline .rubocop.yml:46]: @raul-gracia this seems like a sensible cop. It is a good practice in general.

File: spec/queries/route_breakdown_query_spec.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+
+RSpec.describe RouteBreakdownQuery, type: :model do
+  describe "#call" do
+    let!(:applicant1) { create(:applicant, :teacher) }
```
Comment (pmanrubia) [inline spec/queries/route_breakdown_query_spec.rb:5]: [Following Rubocop](https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/87/files#diff-4f894049af3375c2bd4e608f546f8d4a0eed95464efcdea850993200db9fef5cR77) it seems that `let!` is not needed. 

Why not just:

```ruby
before do
  create :applicant, :teacher
  create:applicant, :salaried_trainee
  create :applicant, :teacher
end
```

We are not making use of `applicant`s variables.

File: spec/models/kpis_spec.rb [hunk 3]
```diff
@@ -5,9 +5,25 @@
...
-      stats = described_class.new
+      kpis = described_class.new
 
-      expect(stats.total_applications).to eq(5)
+      expect(kpis.total_applications).to eq(5)
+    end
+  end
+
+  describe "#route_breakdown" do
+    it "delegates to RouteBreakdownQuery" do
```
Comment (pmanrubia) [inline spec/models/kpis_spec.rb:15]: I would remove this test, as it is mocking the implementation details. 
The unit test for the Query object should be enough alongside the system test.

File: spec/queries/route_breakdown_query_spec.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    let!(:applicant2) { create(:applicant, :salaried_trainee) }
+    let!(:applicant3) { create(:applicant, :teacher) }
+
+    it "returns the correct route breakdown" do
```
Comment (pmanrubia) [inline spec/queries/route_breakdown_query_spec.rb:9]: Can you please add an edge case when there are:
- `0` teachers
- `0` salaried_trainees

File: app/views/system_admin/dashboard/show.html.erb [hunk 5]
```diff
@@ -3,4 +3,15 @@
     <h3 class="title">Applications</h3>
     <p class="value"><%= @kpis.total_applications %></p>
   </div>
+  <div class="kpi-widget routes">
```
Comment (pmanrubia) [inline app/views/system_admin/dashboard/show.html.erb:6]: Shall we extract both: applications and routes into partials?

File: .rubocop.yml (hunk 1 above)
Comment (padv2010) [inline .rubocop.yml:46]: @raul-gracia this seems like a sensible cop. It is a good practice in general.

File: spec/queries/route_breakdown_query_spec.rb (hunk 2 above)
Comment (padv2010) [inline spec/queries/route_breakdown_query_spec.rb:5]: [Following Rubocop](https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/87/files#diff-4f894049af3375c2bd4e608f546f8d4a0eed95464efcdea850993200db9fef5cR77) it seems that let! is not needed.

Why not just:

```ruby
before do
  create :applicant, :teacher
  create:applicant, :salaried_trainee
  create :applicant, :teacher
end
```

We are not making use of applicants variables.

File: spec/models/kpis_spec.rb [hunk 6]
```diff
@@ -5,9 +5,25 @@
...
 
-      stats = described_class.new
+      kpis = described_class.new
 
-      expect(stats.total_applications).to eq(5)
+      expect(kpis.total_applications).to eq(5)
+    end
+  end
+
+  describe "#route_breakdown" do
```
Comment (padv2010) [inline spec/models/kpis_spec.rb:14]: I would remove this test, as it is mocking the implementation details.
The unit test for the Query object should be enough alongside the system test.

File: spec/queries/route_breakdown_query_spec.rb (hunk 4 above)
Comment (padv2010) [inline spec/queries/route_breakdown_query_spec.rb:9]: Can you please add an edge case when there are:
- 0 teachers
- 0 salaried_trainees

File: app/views/system_admin/dashboard/show.html.erb (hunk 5 above)
Comment (padv2010) [inline app/views/system_admin/dashboard/show.html.erb:6]: Shall we extract both: applications and routes into partials?

Comment (raul-gracia) [reply app/views/system_admin/dashboard/show.html.erb:6]: I think I will wait until we have more widgets to create partials.
On the original PR I divided them into "breakdowns/times/totals" 
https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/37/files#diff-20d177abe29b4735fd0aac46984c124d4d383c56103b829817e406bb93971036

File: spec/queries/route_breakdown_query_spec.rb (hunk 2 above)
Comment (raul-gracia) [reply spec/queries/route_breakdown_query_spec.rb:5]: Done

File: spec/queries/route_breakdown_query_spec.rb (hunk 4 above)
Comment (raul-gracia) [reply spec/queries/route_breakdown_query_spec.rb:9]: Done

File: .rubocop.yml (hunk 1 above)
Comment (padv2010) [reply .rubocop.yml]: This comment has not been addressed

Comment (raul-gracia) [reply .rubocop.yml]: It was addressed in this commit made yesterday:
https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/87/commits/2879b2062f53aba990637bb7cc0c2ee923bc3352

And it can be seen on the PR diff here: https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/87/files#diff-4f894049af3375c2bd4e608f546f8d4a0eed95464efcdea850993200db9fef5c

Comment (padv2010) [reply .rubocop.yml]: I still don't think so:

image

with your addition, you are hiding this rubocop issue wich is not a good practice

image

Comment (padv2010) [reply .rubocop.yml]: My previous comment shouldn't have gone in https://github.com/DFE-Digital/get-a-teacher-relocation-payment/pull/87#discussion_r1243633936. All conversations were marked as resolved which contributed of me adding the comment on the wrong conversation.

It is a bit better to let the reviewer to review the comments if possible.

File: spec/queries/route_breakdown_query_spec.rb (hunk 2 above)
Comment (raul-gracia) [reply spec/queries/route_breakdown_query_spec.rb]: After commit https://github.com/DFE-Digital/get-a-teacher-relocation-payment/commit/2879b2062f53aba990637bb7cc0c2ee923bc3352 the `let!` statements weren't there anymore. Therefore there was no need to omit the rubocop rule `RSpec/LetSetup`. It's been removed.

**Comment by raul-gracia** [comment]
@padv2010 let me know if there are any other required changes

**Review by padv2010 — APPROVED** [review]
